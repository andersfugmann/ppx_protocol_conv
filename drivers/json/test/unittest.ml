open Protocol_conv_json
module Driver = struct
  let name = "json"
  let serialize t = Yojson.Safe.pretty_to_string t
  let deserialize t = Yojson.Safe.from_string t
  include Json
  let of_driver_exn = of_json_exn
  let of_driver = of_json
  let to_driver = to_json
  let sexp_of_t : t -> Sexplib.Sexp.t = fun t ->
    Sexplib.Std.sexp_of_string (to_string_hum t)
end

module Unittest = Test.Unittest.Make(Driver)

module Yojson_Driver = struct
  let name = "yojson"
  let serialize t = Yojson.Safe.pretty_to_string t
  let deserialize t = Yojson.Safe.from_string t
    include Json.Yojson
  let of_driver_exn = of_yojson_exn
  let of_driver = of_yojson
  let to_driver = to_yojson
  let sexp_of_t : t -> Sexplib.Sexp.t = fun t ->
    Sexplib.Std.sexp_of_string (to_string_hum t)
end
module Unittest_yojson = Test.Unittest.Make(Yojson_Driver)

(*
module Identity = struct
  open Sexplib.Std
  type v = {int: a; string: b;}
  [@@deriving protocol ~driver:(module Json), sexp]
  type t = { t: Json.t; }
  [@@deriving protocol ~driver:(module Json), sexp]
  let t = { a = 5; b = "s"} |> to_json in
  let t = { t }
  Alcotest.(check (of_pp fmt)) "Deserialize" t (`Assoc y |> of_json_exn);
  Alcotest.(check (of_pp fmt_yojson)) "Serialize" (to_json t) (`Assoc j);


end
*)
let () = Unittest.run ~extra:[Test_attrib.test] ()
let () = Unittest_yojson.run ~extra:[Test_attrib.test] ()
