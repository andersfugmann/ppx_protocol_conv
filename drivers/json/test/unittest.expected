=== json.SingleElem ===
[]
=== json.SingleElem ===
[ 2 ]
=== json.Longarray ===
[ 4, 2, 3, 1 ]
=== json.EmptyInsideRec ===
{ "c": "c", "V": [], "a": "a" }
=== json.SingleInsideRec ===
{ "c": "c", "V": [ 2 ], "a": "a" }
=== json.MultiInsideRec ===
{ "c": "c", "V": [ 4, 2, 3, 1 ], "a": "a" }
=== json.ArrayOfArrays ===
{ "a": [ [ 2, 3 ], [ 4, 5 ] ] }
=== json.ArrayOfArrays2 ===
[ [], [ [], [ 2 ], [ 3, 4 ] ], [ [] ], [ [ 2 ] ] ]
=== json.Tuple ===
[
  [
    10,
    [ 20, 30, 40 ],
    [ "s50", "s60", "s70" ],
    [ [ 100, 200 ], [ 300, 400 ], [ 500, 600 ] ]
  ],
  [
    11,
    [ 21, 31, 41 ],
    [ "s51", "s61", "s71" ],
    [ [ 101, 201 ], [ 301, 401 ], [ 501, 601 ] ]
  ],
  [
    12,
    [ 22, 32, 42 ],
    [ "s52", "s62", "s72" ],
    [ [ 102, 202 ], [ 302, 402 ], [ 502, 602 ] ]
  ],
  [
    13,
    [ 23, 33, 43 ],
    [ "s53", "s63", "s73" ],
    [ [ 103, 203 ], [ 303, 403 ], [ 503, 603 ] ]
  ]
]
=== json.Any ===
{
  "u": { "ua": 7 },
  "v": [ "B", [ 5, 6, 7 ], [ 10, 11, 12 ] ],
  "z": 101,
  "record": { "y": "string", "x": 5 },
  "varray": [
    [ "B", [ 5, 6, 7 ], [ 10, 11, 12 ] ],
    [ "B", [ 5, 6, 7 ], [ 10, 11, 12 ] ],
    [ "B", [ 5, 6, 7 ], [ 10, 11, 12 ] ]
  ],
  "vlist": [
    [ "B", [ 5, 6, 7 ], [ 10, 11, 12 ] ],
    [ "B", [ 5, 6, 7 ], [ 10, 11, 12 ] ],
    [ "B", [ 5, 6, 7 ], [ 10, 11, 12 ] ]
  ],
  "tuple": [ 5, "protocol", false ],
  "intref": 4,
  "intoption": 100,
  "intlist": [ 3, 4, 5 ],
  "bytes": "bytes",
  "string2": "1",
  "string": "string",
  "float": 3.14,
  "nativeint": 20,
  "int64": 10,
  "int32": 5,
  "int": 2,
  "char": "x",
  "bool": true
}
=== json.Record ===
{
  "t_il": [ 1000, 1001, 1002 ],
  "t_tl": [
    [ 100, 101, [ "s100", "s101" ] ],
    [ 110, 111, [ "s110", "s111" ] ],
    [ 120, 121, [ "s120", "s121" ] ],
    [ 130, 131, [ "s130", "s131" ] ],
    [ 140, 141, [ "s140", "s141" ] ]
  ],
  "t_t": [ 100, 101, [ "s100", "s101" ] ],
  "t_i": 1000,
  "t_bl": [],
  "t_b": {
    "b_al": [
      { "a_string": "s7", "a_int": 7 },
      { "a_string": "s8", "a_int": 8 },
      { "a_string": "s9", "a_int": 9 }
    ],
    "b_a": { "a_string": "s6", "a_int": 6 },
    "b_string": "s5",
    "b_int": 5
  },
  "t_al": [
    { "a_string": "s2", "a_int": 2 },
    { "a_string": "s3", "a_int": 3 },
    { "a_string": "s4", "a_int": 4 }
  ],
  "t_a": { "a_string": "s1", "a_int": 1 }
}
=== json.list ===
{ "a": [ 1, 2, 3 ] }
=== json.Lists ===
{
  "l": [
    [ "A", [ 1, 2, 3 ] ],
    [ "B", [ [ 1, 2 ], [ 3, 4, 5 ], [ 2 ] ], [ 3, 1 ], 5 ],
    [ "C", [ 1, 2, 3 ], [ 3, 4, 5 ] ]
  ],
  "c": [ 100, 101, 102, 103 ],
  "b": [ [ 8, 9 ], [ 10, 20, 30, 40 ] ],
  "a": [ [ 1, 2, 3 ], [], [ 10, 20, 30, 40 ], [ 100, 101 ] ]
}
=== json.array ===
{ "a": [ 1, 2, 3 ] }
=== json.EmptyList ===
[]
=== json.SingleElem ===
[ 2 ]
=== json.Longlist ===
[ 4, 3, 2, 1 ]
=== json.EmptyInsideRec ===
{ "c": "c", "V": [], "a": "a" }
=== json.SingleInsideRec ===
{ "c": "c", "V": [ 2 ], "a": "a" }
=== json.MultiInsideRec ===
{ "c": "c", "V": [ 4, 2, 3, 1 ], "a": "a" }
=== json.ListOfLists ===
{ "a": [ [ 2, 3 ], [ 4, 5 ] ] }
=== json.ListOfLists2 ===
[ [], [ [], [ 2 ], [ 3, 4 ] ], [ [] ], [ [ 2 ] ] ]
=== json.Nonrec ===
[ "A", [ "Cons", 4, [ "Cons", 3, "Nil" ] ] ]
=== json.Nonrec2 ===
[ "Cons", 4, [ "Cons", 3, "Nil" ] ]
=== json.None ===
null
=== json.Some None ===
{ "__option": null }
=== json.Some Some None ===
{ "__option": { "__option": null } }
=== json.Some Some Some Unit ===
{ "__option": { "__option": { "__option": { "__option": null } } } }
=== json.simple ===
5
=== json.record ===
{ "a": 5 }
=== json.multiple ===
[ 5, "5", true ]
=== json.reference ===
{ "a": { "a": 5 } }
=== json.recursive ===
{ "c": { "b": { "a": 5 } } }
=== json.Simple ===
[ "A", [ "B", 5 ], [ "C", [ 6, 7 ] ], [ "D", [ 8, 9 ] ] ]
=== json.Tree ===
[ "Node", [ [ "Node", [ "Leaf", 3, "Leaf" ] ], 10, "Leaf" ] ]
=== json.MutualRecursion ===
[
  "T1",
  [ "V", [ "T", [ "V", [ "V1", [ "V1", [ "V1", [ "V0", 5 ] ] ] ] ] ] ]
]
=== json.InsideRec ===
{ "c": "c", "V": "A", "a": "a" }
=== json.RecordList ===
{ "objects": [ { "key": 1 }, { "key": 2 } ] }
=== json.SimpleRecord ===
{
  "HostId": "SDsd",
  "RequestId": "sdfsd",
  "Endpoint": null,
  "Bucket": null,
  "Message": "Message",
  "Code": "Error"
}
=== json.Test_sig ===
{ "x": [ "A", [ [ "A", 7 ], 7, 7, [ "B", 0.7 ], 7 ] ] }
=== json.Test_sig2 ===
[
  1,
  2.0,
  "3.0",
  null,
  [ "A", 1 ],
  { "c": "3.0", "b": 2.0, "a": 1 },
  [ "A", 1 ]
]
=== json.S3 ===
{
  "Contents": [ { "ETag": "Etag", "StorageClass": "STANDARD" } ],
  "Prefix": "prefix"
}
=== json.Types ===
{
  "baz": {
    "y_yd": [ "Variant_two1", 1 ],
    "y_yc": [ "three", [ 100, 200, 300 ] ],
    "y_b": [ "two", [ 10, 20, 30 ] ],
    "y_a": 2
  },
  "bar": "true",
  "foo": 1
}
=== json.Some Some Some true ===
true
=== json.Some Some None ===
{ "__option": { "__option": null } }
=== json.Some None ===
{ "__option": null }
=== json.None ===
null
=== json.Some Some Some true ===
{ "a": true }
=== json.Some Some None ===
{ "a": { "__option": { "__option": null } } }
=== json.Some None ===
{ "a": { "__option": null } }
=== json.None ===
{ "a": null }
=== json.unit option option list option option ===
[ { "__option": { "__option": null } }, { "__option": null }, null ]
=== json.confuse deserialization by using reserved word ===
{ "o": { "option": true } }
=== json.Simple ===
[ "A", [ "B", 5 ], [ "C", 6, 7 ], [ "D", [ 8, 9 ] ] ]
=== json.Tuple ===
[ "A", [ 3, 4 ] ]
=== json.Tree ===
[ "Node", [ "Node", "Leaf", 3, "Leaf" ], 10, "Leaf" ]
=== json.MutualRecursion ===
[
  "T1",
  [ "V", [ "T", [ "V", [ "V1", [ "V1", [ "V1", [ "V0", 5 ] ] ] ] ] ] ]
]
=== json.InsideRec ===
{ "c": "c", "V": "A", "a": "a" }
=== json.InlineRecord ===
[ "A", { "a": "a" } ]
=== json.InlineRecord2 ===
[ "aa", { "b": [ "aa", { "b": [ "B", 5 ], "A": "a" } ], "A": "a" } ]
=== json.Poly ===
[ "aaa", 5 ]
=== json.Option.Ok ===
[ "Ok", 2 ]
=== json.Option.Error ===
[ "Error", "Error string" ]
